plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/FloatDelay.hs

```diff
@@ -0,0 +1,111 @@
+{-# LANGUAGE LambdaCase #-}
+
+module UntypedPlutusCore.Transform.FloatDelay (floatDelay) where
+
+import PlutusCore qualified as PLC
+import UntypedPlutusCore.Core
+
+import Control.Lens
+import Control.Monad
+import Control.Monad.Trans.Writer.CPS
+import Data.Map (Map)
+import Data.Map qualified as Map
+import Data.Set (Set)
+import Data.Set qualified as Set
+
+{- | The Float Delay optimization floats `Delay` from arguments into function bodies,
+if possible. It turns @(\n -> ...Force n...Force n...) (Delay arg)@ into
+@(\n -> ...Force (Delay n)...Force (Delay n)...) arg@.
+
+The above transformation is performed if:
+
+    * All occurrences of @arg@ are under @Force@.
+
+    * @arg@ is essentially work-free.
+
+This achieves a similar effect to Plutonomy's "Split Delay" optimization. The difference
+is that Split Delay simply splits the @Delay@ argument into multiple arguments, turning the
+above example into @(\m -> (\n -> ...Force n...Force n...) (Delay m)) arg@, and then relies
+on other optimizations to simplify it further. Specifically, once the inliner inlines
+@Delay m@, it will be identical to the result of Float Delay.
+
+The advantages of Float Delay are:
+
+    * It is very simple to implement.
+
+    * It doesn't rely on the inliner. In this example, Split Delay relies on the inliner to
+      inline @Delay m@, but there's no guarantee that the inliner will do so, because inlining
+      it may increase the program size.
+
+    * Because Force Delay requires that all occurrences of @arg@ are under @Force@, it
+      guarantees to not increase the size or the cost of the program. This is not the case
+      with Split Delay: in this example, if the occurrences of @n@ are not under @Force@,
+      then Split Delay may increase the size of the program, regardless of whether or not
+      @Delay m@ is inlined. If @Delay m@ is not inlined, then it will also increase the
+      cost of the program, due to the additional application.
```

Comment:
Hmm, although maybe this is also assuming the inliner is stupid: given that the body is big, it probably shouldn't have inlined in step 2, in which case we _would_ do single-occurrence inlining at the end and basically just reverse the original split.

So I remain unconvinced that the combo of split-delay and a sufficiently smart inliner wouldn't be better.

Reply:
> the combo of split-delay and a sufficiently smart inliner

But why make the inliner so smart, and split-delay so dumb? I don't think some super complex inlining heuristics that is aware of what every other optimization does sounds appealing to me.

Also, we've already seen that this idea can be problematic: when inlining fully-saturated functions, we don't do beta-reduction, but simply replace the variable with RHS, and then rely on further inlining to substitute the arguments. But the further inlining may or may not happen. I think it would be better to do everything in one fell swoop.

Reply:
It means that things compose well. If we always make decisions about when to inline variables in _one_ place (the inliner), then no matter how we arrive at such a program, the inliner can always handle it. Whereas with this optimization, if we happen to end up with a program like `let x = delay v in ... force x ...` the inliner won't handle it.